cupy/fft/cache.pyx

```diff
@@ -0,0 +1,600 @@
+        cdef _Node t = self.tail
+        cdef _Node p = t.prev
+        p.next = node
+        t.prev = node
+        node.prev = p
+        node.next = t
+        self.count += 1
+
+
+#####################################################################
+# cuFFT plan cache                                                  #
+#####################################################################
+
+cdef class PlanCache:
+    """A per-thread, per-device, least recently used (LRU) cache for cuFFT
+    plans.
+
+    Args:
+        size (int): The number of plans that the cache can accommodate. The
+            default is 16. Setting this to ``-1`` will make this limit ignored.
+        memsize (int): The amount of GPU memory, in bytes, that the plans in
+            the cache will use for their work areas. Default is ``-1``, meaning
+            it is unlimited.
+        dev (int): The ID of the device that the cache targets.
+
+    .. note::
+        1. By setting either ``size`` to ``0`` (by calling :meth:`set_size`) or
+           ``memsize`` to ``0`` (by calling :meth:`set_memsize`), the cache is
+           disabled, and any operation is no-op. To re-enable it, simply set
+           a nonzero ``size`` and/or ``memsize``.
+
+        2. This class can be instantiated by users, but it is discouraged.
+           Instead, we expect the following canonical usage pattern to
+           retrieve a handle to the cache through :func:`get_plan_cache`:
+
+           .. code-block:: python
+
+               from cupy.cuda import Device
+               from cupy.fft.config import get_plan_cache
+
+               # get the cache for device n
+               with Device(n):
+                   cache = get_plan_cache()
+                   cache.set_size(0)  # disable the cache
+
+           In particular, the cache for device n should be manipulated under
+           the device n's context.
+
+        3. This class is thread-safe since by default it is created on a
+           per-thread basis. When starting a new thread, a new cache is not
+           initialized until :func:`~cupy.fft.config.get_plan_cache` is
+           called or when the constructor is manually invoked.
+
+        4. For multi-GPU plans, the plan will be added to each participating
+           GPU's cache. Upon removal (by any of the caches), the plan will
+           be removed from each participating GPU's cache.
+
+        5. This cache supports the iterator protocol, and returns a 2-tuple:
+           ``(key, node)`` starting from the most recently used plan.
+
+    """
+    # total number of plans, regardless of plan type
+    # -1: unlimited/ignored, cache size is restricted by "memsize"
+    # 0: disable cache
+    cdef Py_ssize_t size
+
+    # current number of cached plans
+    cdef Py_ssize_t curr_size
+
+    # total amount of memory for all of the cached plans
+    # -1: unlimited/ignored, cache size is restricted by "size"
+    # 0: disable cache
+    cdef Py_ssize_t memsize
```

Comment:
This has a different meaning in the SciPy PR. There it is the max size of a _single_ plan, rather than all plans together. So, for example if you have one really large transform it will evict the entire cache here, even if the rest are small and frequently used transforms. Although having limits for both could make sense.

Reply:
Thanks for the heads-up, Peter! I think in the pocketfft case the plan sizes are much smaller and predictable, so a per-plan cap makes better sense. The problem with cuFFT is that we could have super large plans (see https://github.com/cupy/cupy/issues/3777#issuecomment-676561525), so I think it's better to have a cap for all plans, otherwise we might run into many "small" plans, each is say 1 GB, eating up all of the GPU memory.

Reply:
But we do need to revise some of the APIs currently exposed to `cupy.fft.config` when working on the SciPy compilance (next PR). Currently they are marked experimental so it's OK to leave them as is.

Reply:
Well the absolute upper bound is is `cache_size * maxmemsize`. Although your thread-local caches does complicate that quite a bit.

Reply:
Yeah, and we also have multi-GPU plans that make it even more complicated 😉 Plus, `cache_size` could be unlimited (=-1).

Reply:
Well yes, an unlimited `cache_size` doesn't make sense without a total memory limit.